### PR TITLE
gh-103092: Test `_ctypes` type hierarchy and features

### PR DESCRIPTION
**Pull request by aisk**
This change is related to https://github.com/python/cpython/pull/113620#pullrequestreview-1801529989, to test the compatibility before and after it.

All classes with metatypes (which have changed in #113620) are tested, plus `CField`. Other classes are hard to test (can't be accessed in Python easily), or Windows related (I'm not using Windows now), I think it's better to add tests to test them in another PR.

@erlend-aasland

* Issue: gh-103092

### CODE REVIEW

File: Lib/test/test_ctypes/test_struct_fields.py [hunk 1]
```diff
@@ -55,11 +55,21 @@ class X(Structure):
...
+    def test_cfield_instantiation(self):
         class X(Structure):
             _fields_ = [("x", c_int)]
         CField = type(X.x)
-        self.assertRaises(TypeError, CField)
+        self.assertRaisesRegex(TypeError,
+                               "cannot create '_ctypes.CField' instances",
+                               CField)
+
+    def test_cfield_immutability(self):
```
Comment (aisk) [inline Lib/test/test_ctypes/test_struct_fields.py]: `CField` is a basic object (inherif from `object`, no `metaclass`), so not add the inherif test.

File: Lib/test/test_ctypes/test_unions.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        self.assertEqual(type(UnionType), type)
+
+    def test_instantiation(self):
+        with self.assertRaisesRegex(TypeError, "abstract class"):
+            Union()
+
+        UnionType("Foo", (Union,), {})
+
+    def test_immutability(self):
+        Union.foo = "bar"
```
Comment (aisk) [inline Lib/test/test_ctypes/test_unions.py]: To my superise, `Union` is mutable now.

File: Lib/test/test_ctypes/test_arrays.py [hunk 3]
```diff
@@ -23,6 +27,25 @@ def ARRAY(*args):
...
+        self.assertEqual(Array.mro(), [Array, _CData, object])
+
+        self.assertEqual(ArrayType.__name__, "PyCArrayType")
+        self.assertEqual(type(ArrayType), type)
+
+    def test_instantiation(self):
+        with self.assertRaisesRegex(TypeError, "abstract class"):
+            Array()
+
+        ArrayType("Foo", (Array,), {"_length_": 1, "_type_": c_int})
```
Comment (erlend-aasland) [inline Lib/test/test_ctypes/test_arrays.py]: IIRC, we have a test helper in `test.support` for this. `git grep disallow_instantiation` (from the top of my head).

File: Lib/test/test_ctypes/test_arrays.py [hunk 4]
```diff
@@ -14,6 +14,10 @@
 formats = c_byte, c_ubyte, c_short, c_ushort, c_int, c_uint, \
           c_long, c_ulonglong, c_float, c_double, c_longdouble
 
+ArrayType = type(Array)
+# _CData is not exported to Python, so we have to access it from __base__.
+_CData = Structure.__base__
+
```
Comment (erlend-aasland) [inline Lib/test/test_ctypes/test_arrays.py]: How about instead adding a new `TestCase` for the type tests, and move these globals into that class? What do you think?

File: Lib/test/test_ctypes/test_arrays.py (hunk 3 above)
Comment (aisk) [reply Lib/test/test_ctypes/test_arrays.py]: I had taken a look at the `disallow_instantion`, but found that it assumes the error message must have the exact format https://github.com/python/cpython/blob/f19b93fce04fb0bc9b59071915a6aa6b01860d8a/Lib/test/support/__init__.py#L2083.

I think the error is raised by Python runtime, if the `Py_TPFLAGS_DISALLOW_INSTANTIATION` flag is set, so the error message has a fixed format.

In our situation, the error message is "abstract class", I found it's raised in the type's custom `__new__` method, so whether we set the `Py_TPFLAGS_DISALLOW_INSTANTIATION` flag or not, the error will be raised anyway.

So can we use another method to check we kept the same behavior after the change, like check the type's `__flags__`
 field, which is just the `tp_flags` slot in C? The mutability check can be done this way too.

File: Lib/test/test_ctypes/test_arrays.py (hunk 4 above)
Comment (aisk) [reply Lib/test/test_ctypes/test_arrays.py]: I'm not sure if I fully understand what you mean. Did you mean to add a new test file and group all the related tests in it?

File: Lib/test/test_ctypes/test_arrays.py (hunk 3 above)
Comment (erlend-aasland) [reply Lib/test/test_ctypes/test_arrays.py]: Using `__flags__` is nice; I wonder why we didn't do that in the first place 😆 Perhaps we should improve `test.support.disallow_instantiation` to check `__flags__` as well!

Anyway, there is precedent for your suggestion in `Lib/test/test_call.py`. Go ahead!

File: Lib/test/test_ctypes/test_arrays.py (hunk 4 above)
Comment (erlend-aasland) [reply Lib/test/test_ctypes/test_arrays.py]: Sorry for being unclear; yes, that's what I meant. That would also not pollute the global namespace in this file. I'm not sure it is worth it; what you've got is also fine.

File: Lib/test/test_ctypes/test_unions.py (hunk 2 above)
Comment (erlend-aasland) [reply Lib/test/test_ctypes/test_unions.py]: Was it not in earlier Python versions?

File: Lib/test/test_ctypes/test_struct_fields.py (hunk 1 above)
Comment (erlend-aasland) [reply Lib/test/test_ctypes/test_struct_fields.py]: Well, you can test that :)

File: Lib/test/test_ctypes/test_unions.py (hunk 2 above)
Comment (aisk) [reply Lib/test/test_ctypes/test_unions.py]: Yes, testing on 3.10, it's also mutable. I haven't dug too much into it, maybe it's a bug, but I think for now we should keep the behavior?

File: Lib/test/test_ctypes/test_arrays.py (hunk 4 above)
Comment (aisk) [reply Lib/test/test_ctypes/test_arrays.py]: If the concern is the global namespace pollution, I think we can add a `support.py` like other test modules to create the `_CData` and other `XxxTypes`.

File: Lib/test/test_ctypes/test_unions.py (hunk 2 above)
Comment (erlend-aasland) [reply Lib/test/test_ctypes/test_unions.py]: Yes, we aim to keep the current behaviour.

File: Lib/test/test_ctypes/test_arrays.py (hunk 4 above)
Comment (erlend-aasland) [reply Lib/test/test_ctypes/test_arrays.py]: Sounds good to me!

File: Lib/test/test_ctypes/support.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+PyCStructType = type(Structure)
+UnionType = type(Union)
+PyCPointerType = type(_Pointer)
+PyCArrayType = type(Array)
+PyCSimpleType = type(_SimpleCData)
+PyCFuncPtrType = type(CFuncPtr)
+
+# type flags
+PY_TPFLAGS_DISALLOW_INSTANTIATION = 1 << 7
+PY_TPFLAGS_IMMUTABLETYPE = 1 << 8
```
Comment (erlend-aasland) [inline Lib/test/test_ctypes/support.py]:
```suggestion
Py_TPFLAGS_DISALLOW_INSTANTIATION = 1 << 7
Py_TPFLAGS_IMMUTABLETYPE = 1 << 8
```

File: Lib/test/test_ctypes/test_arrays.py [hunk 6]
```diff
@@ -23,6 +25,19 @@ def ARRAY(*args):
...
+        self.assertEqual(PyCArrayType.__name__, "PyCArrayType")
+        self.assertEqual(type(PyCArrayType), type)
+
+    def test_type_flags(self):
+        self.assertTrue(Array.__flags__ & PY_TPFLAGS_IMMUTABLETYPE)
+        self.assertFalse(Array.__flags__ & PY_TPFLAGS_DISALLOW_INSTANTIATION)
+
+        self.assertTrue(PyCArrayType.__flags__ & PY_TPFLAGS_IMMUTABLETYPE)
+        self.assertFalse(PyCArrayType.__flags__ & PY_TPFLAGS_DISALLOW_INSTANTIATION)
```
Comment (erlend-aasland) [inline Lib/test/test_ctypes/test_arrays.py]: You _could_ use `subTest` here, but this is also fine; up to you :)

File: Lib/test/test_ctypes/test_struct_fields.py [hunk 7]
```diff
@@ -12,6 +13,9 @@ class StructFieldsTestCase(unittest.TestCase):
     # 4. The type is subclassed
     #
     # When they are finalized, assigning _fields_ is no longer allowed.
+    class X(Structure):
+        _fields_ = [("x", c_int)]
+    CField = type(X.x)
```
Comment (erlend-aasland) [inline Lib/test/test_ctypes/test_struct_fields.py]: Should this also be placed in `support.py`?

**Comment by aisk** [comment]
Thanks for the review. All issue which comments above should be fixed now.

And renamed the `support.py` to `_support.py`, or it will cause an ambiguous import when running with `-R :` argument:

```
Traceback (most recent call last):
  File "C:\Users\xxxxx\Source\cpython\Lib\unittest\loader.py", line 113, in loadTestsFromModule
    return load_tests(self, tests, pattern)
           ~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\xxxxx\Source\cpython\Lib\test\test_ctypes\__init__.py", line 10, in load_tests
    return support.load_package_tests(os.path.dirname(__file__), *args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: module 'test.test_ctypes.support' has no attribute 'load_package_tests'
```

**Comment by aisk** [comment]
Lots of CIs have failed, seems to be related to GitHub's internal error, and this page often returns 500. I think we should re-run the CI if GitHub fixes the server error.

File: Lib/test/test_ctypes/test_arrays.py [hunk 8]
```diff
@@ -23,6 +25,21 @@ def ARRAY(*args):
...
+        self.assertEqual(type(PyCArrayType), type)
+
+    def test_type_flags(self):
+        with self.subTest(cls=Array):
+            self.assertTrue(Array.__flags__ & Py_TPFLAGS_IMMUTABLETYPE)
+            self.assertFalse(Array.__flags__ & Py_TPFLAGS_DISALLOW_INSTANTIATION)
+
+        with self.subTest(cls=PyCArrayType):
+            self.assertTrue(PyCArrayType.__flags__ & Py_TPFLAGS_IMMUTABLETYPE)
+            self.assertFalse(PyCArrayType.__flags__ & Py_TPFLAGS_DISALLOW_INSTANTIATION)
```
Comment (erlend-aasland) [inline Lib/test/test_ctypes/test_arrays.py]: Sorry for not being clearer; the `subTest` usage would be:

```suggestion
        for cls in Array, PyCArrayType:
            with self.subTest(cls=cls):
                self.assertTrue(cls.__flags__ & Py_TPFLAGS_IMMUTABLETYPE)
                self.assertFalse(cls.__flags__ & Py_TPFLAGS_DISALLOW_INSTANTIATION)
```

Ditto for the other tests.

Comment (aisk) [reply Lib/test/test_ctypes/test_arrays.py]: I got the point, updated.

**Review by erlend-aasland — APPROVED** [review]

**Pull request merged**